### PR TITLE
[devbox.json] Add and implement imports field

### PR DESCRIPTION
**Pull request by mikeland73**
## Summary

Adds and implements `imports` field to devbox.json. Imports follow flake ref specs (but are limited to only path and github for now).

For example:

```json
"imports": [
  "./foobar",
  "path:./foobar2",
  "github:jetpack-io/devbox?dir=somedir",
]
```

Fields from imported configs are added to the environment. The merge algorithm is as follows:

* Imports are evaluated in order first. Base config is evaluated last.
* Slices are concatenated (init hooks, etc)
* maps are merged with later evaluation replacing previous one on conflict.
* Packages are deduped by name and concatenated. (i.e. multiple versions not allowed)

Caveats:

* This can be slow. When using remote imports, every devbox command does 1 or more HTTP requests to load the imports.
* central lock file. All dependencies are stored there. 

TODOs:

* Handle relative paths inside of imported configs
* Cycle detection

## How was it tested?

* Tested importing local devbox.json
* Tested importing github  devbox.json

### CODE REVIEW

File: nix/flake/reflike.go
Comment (gcurtis) [inline nix/flake/reflike.go:1]: I think there's a way we can leverage Nix's ability to fetch flake refs. I'll add it to the review comment.

**Review by gcurtis — COMMENTED** [review]
Instead of adding our own download logic, we could use `nix flake prefetch --json <flake-ref>` to fetch the directory containing the devbox.json. Contrary to the name, it can download non-flake sources:

> This command downloads the source tree denoted by flake reference flake-url. Note that this does not need to be a flake (i.e. it does not have to contain a flake.nix file).

I think this would have a few advantages:

- All flake references work automatically.
- Nix handles caching for us by downloading to the Nix store.
- We get a bunch of info back that would be useful for locking the included config (if we want to eventually implement that).

Using the devbox repo as an example:

```
$ nix flake prefetch --json github:jetpack-io/devbox | jq
{
  "hash": "sha256-7i6hmy1EOAeNW6TXZKoXsBeuRksx3QDrqzatOGXqR1g=",
  "locked": {
    "lastModified": 1708479263,
    "narHash": "sha256-7i6hmy1EOAeNW6TXZKoXsBeuRksx3QDrqzatOGXqR1g=",
    "owner": "jetpack-io",
    "repo": "devbox",
    "rev": "b0d6032376ee21bd9d428358a3046984770e5fd2",
    "type": "github"
  },
  "original": {
    "owner": "jetpack-io",
    "repo": "devbox",
    "type": "github"
  },
  "storePath": "/nix/store/psjz43j55y9qx2sbpr035qywmrj42031-source"
}
$ cat /nix/store/psjz43j55y9qx2sbpr035qywmrj42031-source/devbox.json
{
  "name":        "devbox",
  "description": "Instant, easy, and predictable development environments",
...
```

We can also prefetch local paths, but they have to be in the URL form (`path:./foo` not `./foo`). The `flake.Ref.String()` method can already handle converting those though.

**Comment by gcurtis** [comment]
It can download individual files if you transform the `github` flake reference to a `file` one. For example, `nix flake prefetch --json file+https://raw.githubusercontent.com/jetpack-io/devbox/main/devbox.json`.

That said, I think that downloading the full tree has some advantages.

- You can have a devbox.json that references files within its own repo. For example, something like axiom's Go linting and tidy scripts.
- You can have common plugin configs, such as an `nginx.conf`.
- It aligns with how flake references work, so we don't need to add another syntax.

We don't need to worry about caching if we rely on Nix doing it. After the initial download, `nix flake prefetch` returns instantly and skips the download. We could always make it even faster by incorporating it into the Devbox cache, but it's fast enough that we don't need to think about it for the first implementation.

File: nix/flake/reflike.go
Comment (gcurtis) [reply nix/flake/reflike.go:1]: If you end up leaving in the Devbox fetching logic, could the reflike code be moved to another package? Just so we can keep nix/flake specific to nix.

Comment (mikeland73) [reply nix/flake/reflike.go:1]: > If you end up leaving in the Devbox fetching logic, could the reflike code be moved to another package? Just so we can keep nix/flake specific to nix

Yep, for sure. I was thinking of this, but didn't end up doing it. I can do as part of this PR.

**Review by savil — CHANGES_REQUESTED** [review]
putting back on your queue, since I think we're no longer doing this. Please re-request review if needed.

**Comment by mikeland73** [comment]
Closing in favor of https://github.com/jetpack-io/devbox/pull/1850

**Pull request closed without merging**